sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -1331,6 +1331,29 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val DESERIALIZATION_FACTOR_CALC_ENABLED =
+    buildConf("spark.sql.statistics.deserFactor.calc.enabled")
+      .doc("Enables the calculation of the deserialization factor as a table statistic. " +
+        "This factor is intended to be calculated for columnar storage formats as a ratio of " +
+        "actual data size to raw file size but currently Spark calculates this only for the ORC " +
+        "format. Spark uses this ratio is to scale up the estimated size, which leads to " +
+        "better estimate of in-memory data size and improves the query optimization (i.e., join " +
+        "strategy). In case of partitioned table the maximum of these factors is taken. " +
+        "Spark stores this factor in the meta store and reuses it so the table " +
+        "can grow without having to recompute this statistic. " +
+        "The stored factor can be removed only by a TRUNCATE or a DROP table so even a " +
+        "subsequent ANALYZE TABLE where the calculation is disabled keeps the old value.")
+      .booleanConf
+      .createWithDefault(false)
+
+  val DESERIALIZATION_FACTOR_EXTRA_DISTORTION =
+    buildConf("spark.sql.statistics.deserFactor.distortion")
```

Comment:
Sorry if I rushed to read but why do we need this configuration?

Reply:
This extra factor provides some extra safety to avoid the OOM as the deserialisation factor on its own just a good estimate (much better than the file size, but not still not the exact size of the map which built for broadcast join).

Reply:
Hm .. so .. does this PR propose below?

1. get the size of file in ORC and use it as stats
2. one configuration to control the ratio of the size

If that's correct, looks like we should better separate PRs.
Also, do we have some perf numbers? I have some doubts per https://github.com/apache/spark/pull/26016#discussion_r338386395 . It reminds me of `mergeSchema` in Parquet, which brings considerable performance penalty.

Reply:
> Hm, wouldn't this require the driver side to touch and read every file?

Having just experienced an issue where Spark was attempting to infer the schema on various parquet tables, each with thousands of partitions, I can see that touching all the files could be painful (and just not acceptable to some users).

The change to make the table property settable by the user alleviates this somewhat, in that the user can set a SWAG or run an external tool that only looks at a subset of files and/or partitions to calculate a reasonable ratio.

Reply:
Yes the calculation can be switched on/off and I am about to prove the deserialization factor calculation for a single file is not a problem (by extending `OrcReadBenchmark` with a new case).  

As the whole change is not so huge and both are very much related I would like to keep this PR together.

Reply:
Benchmark result:
```
[info] Running benchmark: DeserFactor calculation with 100 columns
[info]   Running case: Native ORC (no deserFactor calc)
[info]   Stopped after 2 iterations, 2454 ms
[info]   Running case: Native ORC (with deserFactor calc)
[info]   Stopped after 2 iterations, 2001 ms
[info]
[info] OpenJDK 64-Bit Server VM 1.8.0_191-8u191-b12-0ubuntu0.18.04.1-b12 on Linux 4.15.0-1021-aws
[info] Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
[info] DeserFactor calculation with 100 columns:  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] Native ORC (no deserFactor calc)                   1098           1227         183          9.6         104.7       1.0X
[info] Native ORC (with deserFactor calc)                 1000           1001           1         10.5          95.3       1.1X
[info]
[info] Running benchmark: DeserFactor calculation with 200 columns
[info]   Running case: Native ORC (no deserFactor calc)
[info]   Stopped after 2 iterations, 8055 ms
[info]   Running case: Native ORC (with deserFactor calc)
[info]   Stopped after 2 iterations, 7462 ms
[info]
[info] OpenJDK 64-Bit Server VM 1.8.0_191-8u191-b12-0ubuntu0.18.04.1-b12 on Linux 4.15.0-1021-aws
[info] Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
[info] DeserFactor calculation with 200 columns:  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] Native ORC (no deserFactor calc)                   3844           4028         260          2.7         366.6       1.0X
[info] Native ORC (with deserFactor calc)                 3690           3731          59          2.8         351.9       1.0X
[info]
[info] Running benchmark: DeserFactor calculation with 300 columns
[info]   Running case: Native ORC (no deserFactor calc)
[info]   Stopped after 2 iterations, 22299 ms
[info]   Running case: Native ORC (with deserFactor calc)
[info]   Stopped after 2 iterations, 21845 ms
[info]
[info] OpenJDK 64-Bit Server VM 1.8.0_191-8u191-b12-0ubuntu0.18.04.1-b12 on Linux 4.15.0-1021-aws
[info] Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
[info] DeserFactor calculation with 300 columns:  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] Native ORC (no deserFactor calc)                  11094          11150          80          0.9        1058.0       1.0X
[info] Native ORC (with deserFactor calc)                10916          10923           9          1.0        1041.1       1.0X
[info]
[success] Total time: 931 s, completed Dec 6, 2019 8:47:34 AM
```

Reply:
Resolved with 3ee9be61b504887972a8cc09ca7ce7419e196fa2